futures-util/src/future/join.rs

```diff
@@ -1,81 +1,59 @@
-#![allow(non_snake_case)]
-
-use super::assert_future;
-use crate::future::{maybe_done, MaybeDone};
+use crate::future::{assert_future, maybe_done, MaybeDone};
 use core::fmt;
 use core::pin::Pin;
 use futures_core::future::{FusedFuture, Future};
 use futures_core::task::{Context, Poll};
 use pin_project_lite::pin_project;
 
-macro_rules! generate {
-    ($(
-        $(#[$doc:meta])*
-        ($Join:ident, <$($Fut:ident),*>),
-    )*) => ($(
-        pin_project! {
-            $(#[$doc])*
-            #[must_use = "futures do nothing unless you `.await` or poll them"]
-            pub struct $Join<$($Fut: Future),*> {
-                $(#[pin] $Fut: MaybeDone<$Fut>,)*
-            }
-        }
+pin_project! {
+    /// Future for the [`join`](super::FutureExt::join) method.
+    #[must_use = "futures do nothing unless you `.await` or poll them"]
+    pub struct Join<Fut1, Fut2> where Fut1: Future, Fut2: Future {
+        #[pin] fut1: MaybeDone<Fut1>,
+        #[pin] fut2: MaybeDone<Fut2>,
+    }
+}
 
-        impl<$($Fut),*> fmt::Debug for $Join<$($Fut),*>
-        where
-            $(
-                $Fut: Future + fmt::Debug,
-                $Fut::Output: fmt::Debug,
-            )*
-        {
-            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-                f.debug_struct(stringify!($Join))
-                    $(.field(stringify!($Fut), &self.$Fut))*
-                    .finish()
-            }
-        }
+impl<Fut1: Future, Fut2: Future> Join<Fut1, Fut2> {
+    pub(crate) fn new(fut1: Fut1, fut2: Fut2) -> Self {
+        Self { fut1: maybe_done(fut1), fut2: maybe_done(fut2) }
+    }
+}
 
-        impl<$($Fut: Future),*> $Join<$($Fut),*> {
-            fn new($($Fut: $Fut),*) -> Self {
-                Self {
-                    $($Fut: maybe_done($Fut)),*
-                }
-            }
-        }
+impl<Fut1, Fut2> fmt::Debug for Join<Fut1, Fut2>
+where
+    Fut1: Future + fmt::Debug,
+    Fut1::Output: fmt::Debug,
+    Fut2: Future + fmt::Debug,
+    Fut2::Output: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Join").field("fut1", &self.fut1).field("fut2", &self.fut2).finish()
+    }
+}
```

Comment:
I'm not sure why the previous code didn't do it, but this can probably be replaced by `#[derive(Debug)]`.

Reply:
The `Debug` macro can't handle associated type trait bounds:
```rust
error[E0277]: `<Fut1 as Future>::Output` doesn't implement `Debug`
  --> futures-util/src/future/join.rs:11:14
   |
11 |     #[derive(Debug)]
   |              ^^^^^ `<Fut1 as Future>::Output` cannot be formatted using `{:?}` because it doesn't implement `Debug`
```

Reply:
Oh, sorry! I missed that problem.